packages/parser/src/ethereum/index.ts

```diff
@@ -0,0 +1,180 @@
+import { BigNumber } from 'bignumber.js'
+import { ethers } from 'ethers'
+import { Tx } from '@shapeshiftoss/blockbook'
+import { Tx as ParseTx, TxSpecific as ParseTxSpecific, Token, TransferType, Transfer } from '../types'
+import { InternalTx, Network } from './types'
+import { getSigHash } from './utils'
+import * as multiSig from './multiSig'
+import * as thor from './thor'
+import * as uniV2 from './uniV2'
+import * as zrx from './zrx'
+
+export interface TransactionParserArgs {
+  network?: Network
+  midgardUrl: string
+  rpcUrl: string
+}
+
+export class TransactionParser {
+  network: Network
+
+  private thor: thor.Parser
+  private uniV2: uniV2.Parser
+  private zrx: zrx.Parser
+
+  constructor(args: TransactionParserArgs) {
+    const provider = new ethers.providers.JsonRpcProvider(args.rpcUrl)
+
+    this.network = args.network ?? Network.Mainnet
+
+    this.thor = new thor.Parser({ network: this.network, midgardUrl: args.midgardUrl })
+    this.uniV2 = new uniV2.Parser({ network: this.network, provider })
+    this.zrx = new zrx.Parser()
+  }
+
+  // return any addresses that can be detected
+  getInternalAddress(inputData: string): string | undefined {
+    switch (getSigHash(inputData)) {
+      case this.thor.transferOutSigHash:
+        return this.thor.getInternalAddress(inputData)
+      case multiSig.SENDMULTISIG_SIG_HASH:
+        return multiSig.getInternalAddress(inputData)
+      default:
+        return
+    }
+  }
+
+  async parse(tx: Tx, address: string, internalTxs?: Array<InternalTx>): Promise<ParseTx> {
+    const sendAddress = tx.vin[0].addresses?.[0] ?? ''
+    const receiveAddress = tx.vout[0].addresses?.[0] ?? ''
```

Comment:
Do we need any level of validation to consider a transaction valid?

Reply:
Can always log if there is a case where we don't have addresses to look into further. Blockbook is going to send back all transactions it finds in this format or return an error.